### PR TITLE
Give library users direct access to the Inputstream

### PR DESCRIPTION
**Pull request by loosebazooka**
This removes our restriction of offering the user a line at a time, see DeployExample for more.

### CODE REVIEW

File: src/main/java/com/google/cloud/tools/appengine/experimental/deploy/DeployExample.java [hunk 1]
```diff
@@ -38,22 +41,36 @@ public static void main(String[] args) {
...
     config.setDeployables(Collections.singletonList(new File("/tmp/app.yaml")));
 
-    // the current implementation doesn't have a good hook in to the autodetection of the
-    // cloud sdk.
     AppEngineRequestFactory requestFactory = AppEngineRequests.newRequestFactoryBuilder()
         //.cloudSdk(Paths.get("/path/to/cloudsk"))
         // or explicitly tell it to look for it
-        .cloudSdk(
-            new PathResolver().getCloudSdkPath()) //<<-- path resolver would go into the public API
+        .cloudSdk(new PathResolver().getCloudSdkPath()) //<<-- path resolver is in the public API
```
Comment (elharo) [inline src/main/java/com/google/cloud/tools/appengine/experimental/deploy/DeployExample.java]: I don't think this comment is needed

File: src/main/java/com/google/cloud/tools/appengine/experimental/deploy/DeployExample.java [hunk 2]
```diff
@@ -38,22 +41,36 @@ public static void main(String[] args) {
...
+
+    // perhaps we could offer some abstraction of this, but essentially in this modification
+    // we're giving the user straight accesss to the input stream (stderr from gcloud)
+    Thread thread = new Thread(new Runnable() {
+      @Override
+      public void run() {
+        try (Reader r = new InputStreamReader(is)) {
+          int ch;
+          while ((ch = r.read()) != -1) {
+            System.out.print((char) ch);
```
Comment (elharo) [inline src/main/java/com/google/cloud/tools/appengine/experimental/deploy/DeployExample.java]: libraries usually don't write to System.out

File: src/main/java/com/google/cloud/tools/appengine/experimental/internal/cloudsdk/CloudSdkRequestFuture.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import com.google.cloud.tools.appengine.experimental.AppEngineRequestFuture;
+import com.google.cloud.tools.appengine.experimental.internal.process.CliProcessManager;
+
+import java.io.InputStream;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public class CloudSdkRequestFuture<T> implements AppEngineRequestFuture<T> {
```
Comment (elharo) [inline src/main/java/com/google/cloud/tools/appengine/experimental/internal/cloudsdk/CloudSdkRequestFuture.java]: this class needs a test

File: src/main/java/com/google/cloud/tools/appengine/experimental/deploy/DeployExample.java [hunk 4]
```diff
@@ -19,13 +19,16 @@
...
 import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
 import java.util.Collections;
 import java.util.concurrent.ExecutionException;
-import java.util.concurrent.Future;
 
 public class DeployExample {
```
Comment (elharo) [inline src/main/java/com/google/cloud/tools/appengine/experimental/deploy/DeployExample.java]: What's the purpose of this class? If it's example code it doesn't really belong in this package. It's more of a tutorial thing.

File: src/main/java/com/google/cloud/tools/appengine/experimental/internal/cloudsdk/CloudSdkRequest.java [hunk 5]
```diff
@@ -37,8 +35,7 @@
   private final CloudSdkProcessFactory processFactory;
   private final CliProcessManagerProvider<R> processManagerProvider;
   private final StringResultConverter<R> resultConverter;
-  private OutputHandler outputHandler;
-  private boolean mutable = true;
+  private boolean executed = false;
 
   /**
    * Create a new CloudSdkRequest, there are no subclasses, just configure the providers correctly.
```
Comment (elharo) [inline src/main/java/com/google/cloud/tools/appengine/experimental/internal/cloudsdk/CloudSdkRequest.java]: f there are no subclasses this class should be final

File: src/main/java/com/google/cloud/tools/appengine/experimental/internal/cloudsdk/CloudSdkRequestFuture.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2016 Google Inc.
```
Comment (chanseokoh) [inline src/main/java/com/google/cloud/tools/appengine/experimental/internal/cloudsdk/CloudSdkRequestFuture.java]: 2017

File: src/main/java/com/google/cloud/tools/appengine/experimental/deploy/DeployExample.java [hunk 7]
```diff
@@ -38,22 +41,36 @@ public static void main(String[] args) {
...
+    AppEngineRequestFuture<DeployResult> deployFuture =
+        requestFactory.newDeploymentRequest(config).execute();
+    final InputStream is = deployFuture.getInputStream();
+
+    // perhaps we could offer some abstraction of this, but essentially in this modification
+    // we're giving the user straight accesss to the input stream (stderr from gcloud)
+    Thread thread = new Thread(new Runnable() {
+      @Override
+      public void run() {
+        try (Reader r = new InputStreamReader(is)) {
```
Comment (chanseokoh) [inline src/main/java/com/google/cloud/tools/appengine/experimental/deploy/DeployExample.java]: I wonder if we have to specify an encoding.

File: src/main/java/com/google/cloud/tools/appengine/experimental/AppEngineRequest.java [hunk 8]
```diff
@@ -16,15 +16,14 @@
...
-import java.util.concurrent.Future;
+import com.google.cloud.tools.appengine.experimental.internal.cloudsdk.CloudSdkRequestFuture;
+
+import java.io.InputStream;
 
 /**
  * Interface for remote App Engine Requests.
  * @param <T> return request result type
  */
 public interface AppEngineRequest<T> {
```
Comment (chanseokoh) [inline src/main/java/com/google/cloud/tools/appengine/experimental/AppEngineRequest.java]: Question: does a client need parameterization on this class? Looking at the example code below, seems like clients will use `AppEngineRequestFactory`, which returns `AppEngineRequest<DeployResult>`. Are there uses cases where clients want somethings else than `DeployResult`? If so, will client by-pass the factory?

File: src/main/java/com/google/cloud/tools/appengine/experimental/internal/cloudsdk/CloudSdkRequestFuture.java (hunk 3 above)
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/appengine/experimental/internal/cloudsdk/CloudSdkRequestFuture.java]: This class is redundant, I'm taking it out.

File: src/main/java/com/google/cloud/tools/appengine/experimental/internal/cloudsdk/CloudSdkRequest.java (hunk 5 above)
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/appengine/experimental/internal/cloudsdk/CloudSdkRequest.java]: will do.

File: src/main/java/com/google/cloud/tools/appengine/experimental/deploy/DeployExample.java (hunk 7 above)
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/appengine/experimental/deploy/DeployExample.java]: Yeah, this is just an example, I expect to take this class out.

File: src/main/java/com/google/cloud/tools/appengine/experimental/AppEngineRequest.java (hunk 8 above)
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/appengine/experimental/AppEngineRequest.java]: Yeah DeployResult is just a weak wrapper around a string. Potentially for other requests you can return more structured data.

`AppEngineRequestFuture<List<Instances>>` or something might be an example

File: src/main/java/com/google/cloud/tools/appengine/experimental/deploy/DeployExample.java (hunk 4 above)
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/appengine/experimental/deploy/DeployExample.java]: Yeah this has kind of been in there to show us how to use it. It has no necessity for being there. It's not going to be publicly documented until we decide exactly what it is going to be.

File: src/main/java/com/google/cloud/tools/appengine/experimental/deploy/DeployExample.java (hunk 2 above)
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/appengine/experimental/deploy/DeployExample.java]: just a sample.

File: src/main/java/com/google/cloud/tools/appengine/experimental/internal/cloudsdk/CloudSdkRequestFuture.java (hunk 6 above)
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/appengine/experimental/internal/cloudsdk/CloudSdkRequestFuture.java]: this class is gone.

File: src/main/java/com/google/cloud/tools/appengine/experimental/internal/process/CliProcessManager.java [hunk 9]
```diff
@@ -43,28 +43,24 @@
  *
  * @param <T> the process return type
  */
-public class CliProcessManager<T> implements Future<T> {
+public class CliProcessManager<T> implements AppEngineRequestFuture<T> {
```
Comment (patflynn) [inline src/main/java/com/google/cloud/tools/appengine/experimental/internal/process/CliProcessManager.java]: CliCommandRequestProcessor implements ..?

File: src/main/java/com/google/cloud/tools/appengine/experimental/deploy/DeployExample.java (hunk 4 above)
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/appengine/experimental/deploy/DeployExample.java]: moved this into a readme.

**Pull request merged**